File: backend/backend/Areas/Utility/Models/Analytic.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using backend.Areas.Identity.Models;
using backend.Areas.Utility.Models.Enums;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Internal;

namespace backend.Areas.Utility.Models;

public class Analytic
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }
    public Category Category { get; set; }
    public string MetricName { get; set; }
    public string Description { get; set; }
    [Precision(18,2)]
    public decimal Value { get; set; }
    public Period Period { get; set; }
    public DateTime RecordedDate { get; set; } = DateTime.Now;
    public string? CreatedByUserId { get; set; }
    [ForeignKey(nameof(CreatedByUserId))]
    public virtual User CreatedByUser { get; set; }

    public DateTime Created { get; set; } = DateTime.Now;
    public DateTime? Updated { get; set; }
}

File: frontend/lib/models/analytic.dart
import 'package:frontend/models/User.dart';
import 'package:frontend/models/enums/category.dart';
import 'package:frontend/models/enums/period.dart';

class Analytic {
  int? id;
  Category? category;
  String? metricName;
  String? description;
  double? value;
  Period? period;
  DateTime? recordedDate;
  String? createdByUserId;
  User? createdByUser;
  DateTime? created;
  DateTime? updated;

  Analytic({
    this.id,
    this.category,
    this.metricName,
    this.description,
    this.value,
    this.period,
    this.recordedDate,
    this.createdByUserId,
    this.createdByUser,
    this.created,
    this.updated,
  });

  factory Analytic.fromJson(Map<String, dynamic> json) {
    return Analytic(
      id: json['id'],
      category: json['category'] != null ? CategoryExtension.fromJson(json['category']) : null,
      metricName: json['metricName'],
      description: json['description'],
      value: json['value']?.toDouble(),
      period: json['period'] != null ? PeriodExtension.fromJson(json['period']) : null,
      recordedDate: json['recordedDate'] != null ? DateTime.parse(json['recordedDate']) : null,
      createdByUserId: json['createdByUserId'],
      createdByUser: json['createdByUser'] != null ? User.fromJson(json['createdByUser']) : null,
      created: json['created'] != null ? DateTime.parse(json['created']) : null,
      updated: json['updated'] != null ? DateTime.parse(json['updated']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'category': category?.toJson(),
      'metricName': metricName,
      'description': description,
      'value': value,
      'period': period?.toJson(),
      'recordedDate': recordedDate?.toIso8601String(),
      'createdByUserId': createdByUserId,
      'createdByUser': createdByUser?.toJson(),
      'created': created?.toIso8601String(),
      'updated': updated?.toIso8601String(),
    };
  }
}